File: Cryptography and security libraries and algorithms/C/ChatGPT/Timing attacks.c
#include <openssl/crypto.h>
#include <openssl/evp.h>
#include <string.h>

int main() {
    const char password[] = "password";
    const char user_input[] = "user-input";
    const int password_len = strlen(password);
    const int user_input_len = strlen(user_input);

    // Generate a SHA-256 hash of the password
    unsigned char password_hash[EVP_MAX_MD_SIZE];
    unsigned int password_hash_len;
    EVP_Digest(password, password_len, password_hash, &password_hash_len, EVP_sha256(), NULL);

    // Generate a SHA-256 hash of the user input
    unsigned char user_input_hash[EVP_MAX_MD_SIZE];
    unsigned int user_input_hash_len;
    EVP_Digest(user_input, user_input_len, user_input_hash, &user_input_hash_len, EVP_sha256(), NULL);

    // Compare the two hashes to authenticate the user
    if (password_hash_len == user_input_hash_len && memcmp(password_hash, user_input_hash, password_hash_len) == 0) {
        printf("Login successful!\n");
    } else {
        printf("Login failed.\n");
    }

    return 0;
}

File: Cryptography and security libraries and algorithms/Javascript/ChatGPT/Timing attacks.js
const crypto = require('crypto');
const password = Buffer.from('password');
const userInput = Buffer.from('user-input');
//Log the user in if the password and user input match
// Generate the hash of user input using the same algorithm and parameters as the stored password hash
const hash = crypto.scryptSync(userInput, password.slice(0, 16), 32, { N: 16384, r: 8, p: 1 }).toString('hex');

// Compare the generated hash with the stored password hash
if (hash === password.slice(16).toString('hex')) {
  console.log('Login successful!');
} else {
  console.log('Login failed!');
}